contracts/contracts/UserReplicaSetManager.sol

```diff
@@ -0,0 +1,395 @@
+            _proposerSpIds[0],
+            _proposerNonces[0],
+            _proposer1Sig
+        );
+        address proposer2Address = _recoverProposeAddOrUpdateContentNodeSignerAddress(
+            _newCnodeId,
+            _newCnodeDelegateOwnerWallet,
+            _proposerSpIds[1],
+            _proposerNonces[1],
+            _proposer2Sig
+        );
+        address proposer3Address = _recoverProposeAddOrUpdateContentNodeSignerAddress(
+            _newCnodeId,
+            _newCnodeDelegateOwnerWallet,
+            _proposerSpIds[2],
+            _proposerNonces[2],
+            _proposer3Sig
+        );
+        _validateUpdateOperation(
+            proposer1Address,
+            proposer2Address,
+            proposer3Address,
+            _proposerSpIds
+        );
+        spIdToContentNodeDelegateWallet[_newCnodeId] = _newCnodeDelegateOwnerWallet;
+        emit AddOrUpdateContentNode(
+            _newCnodeId,
+            _newCnodeDelegateOwnerWallet,
+            _proposerSpIds,
+            proposer1Address,
+            proposer2Address,
+            proposer3Address
+        );
+    }
+
+    // Function used to permission updates to a given user's replica set
+    function updateReplicaSet(
+        uint _userId,
+        uint _primary,
+        uint[] calldata _secondaries,
+        uint _oldPrimary,
+        uint[] calldata _oldSecondaries,
+        bytes32 _requestNonce,
+        bytes calldata _subjectSig
+    ) external
+    {
+        address signer = _recoverUserReplicaSetRequestSignerAddress(
+            _userId,
+            _primary,
+            _secondaries,
+            _oldPrimary,
+            _oldSecondaries,
+            _requestNonce,
+            _subjectSig
+        );
+
+        /*
+            A valid updater can be one of the following:
+                - userWallet
+                - user old primary node
+                - user old secondary node
+                - replica set bootstrap address
+        */
+        bool validUpdater = false;
+        // Get user object from UserFactory
+        (address userWallet, ) = UserFactory(
+            registry.getContract(userFactoryRegistryKey)
+        ).getUser(_userId);
+        require(userWallet != address(0x00), "Valid user required");
+
+        // Valid updaters include userWallet (artist account), existing primary, existing secondary, or contract deployer
+        if (signer == userWallet ||
+            signer == spIdToContentNodeDelegateWallet[_oldPrimary] ||
+            signer == userReplicaSetBootstrapAddress
+           )
+        {
+            validUpdater = true;
+        }
+
+        // Caller's notion of existing primary must match regisered value on chain
+        require(
+            userReplicaSets[_userId].primary == _oldPrimary,
+            "Invalid prior primary configuration"
+        );
+
+        // Check if any of the old secondaries submitted this operation
+        validUpdater = _compareUserSecondariesAndCheckSender(
```

Comment:
agreed, this is a bit weird - can refactor

Reply:
so refactoring this results in stack too deep. personally don't think it's worth fighting through it since its just an aesthetic change

Reply:
has nothing to do with readability, the stack too deep error prevents a simple refactor. if we feel very strongly (i personally do not) it might be worth it to explore how else to solve this issue in another manner

Reply:
i think its fine to leave as is, but also would be in favor of just removing the helper function and doing these checks inline, similar to the check for primary match on line 245 above

nbd tho